### PR TITLE
HWKAPM-609 Calculate latency for 'multi-consumer' relationships

### PR DESCRIPTION
**Pull request by objectiser**

### CODE REVIEW

**Comment by objectiser** [comment]
@pavolloffay could you review please?

File: processors/communicationdetails-deriver/src/main/java/org/hawkular/apm/processor/communicationdetails/CommunicationDetailsDeriver.java [hunk 1]
```diff
@@ -183,6 +175,41 @@ public CommunicationDetails processOneToOne(String tenantId, Trace item) throws
...
+                latency = calculateTimestampLatency(si, item, consumer);
+            }
+        } else {
+            latency = calculateTimestampLatency(si, item, consumer);
+        }
+
+        return latency;
+    }
+
+    private static long calculateTimestampLatency(SourceInfo si, Trace item, Consumer consumer) {
```
Comment (pavolloffay) [inline processors/communicationdetails-deriver/src/main/java/org/hawkular/apm/processor/communicationdetails/CommunicationDetailsDeriver.java]: `consumer` is used only for logging?

File: processors/communicationdetails-deriver/src/main/java/org/hawkular/apm/processor/communicationdetails/CommunicationDetailsDeriver.java [hunk 2]
```diff
@@ -183,6 +175,41 @@ public CommunicationDetails processOneToOne(String tenantId, Trace item) throws
...
     }
 
+    protected static long calculateLatency(SourceInfo si, Trace item, Consumer consumer) {
+        long latency = 0;
+
+        if (!si.isMultipleConsumers()) {
+            long diff = TimeUnit.MILLISECONDS.convert(si.getDuration() - consumer.getDuration(),
+                    TimeUnit.NANOSECONDS);
+            if (diff > 0) {
+                latency = diff / 2;
```
Comment (pavolloffay) [inline processors/communicationdetails-deriver/src/main/java/org/hawkular/apm/processor/communicationdetails/CommunicationDetailsDeriver.java]: `diff >> 1`

**Comment by pavolloffay** [comment]
Review finished LGTM.

This implementation assumes that request and response latency are the same. However, if the request goes via internet than the request and response latency can significantly differ.

File: processors/communicationdetails-deriver/src/main/java/org/hawkular/apm/processor/communicationdetails/CommunicationDetailsDeriver.java (hunk 2 above)
Comment (pavolloffay) [inline processors/communicationdetails-deriver/src/main/java/org/hawkular/apm/processor/communicationdetails/CommunicationDetailsDeriver.java]: also please add a comment why there is a division.

File: processors/communicationdetails-deriver/src/main/java/org/hawkular/apm/processor/communicationdetails/CommunicationDetailsDeriver.java (hunk 1 above)
Comment (objectiser) [inline processors/communicationdetails-deriver/src/main/java/org/hawkular/apm/processor/communicationdetails/CommunicationDetailsDeriver.java]: yes, but will change to just use the trace instead.

**Comment by objectiser** [comment]
@pavolloffay Thanks!

**Pull request merged**